{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import xarray as xr\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import cartopy\n",
    "import cartopy.io\n",
    "import cartopy.io.img_tiles\n",
    "import cartopy.feature\n",
    "import shapely\n",
    "\n",
    "import functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "skip_ncgen_prcp = True # set 'False' for the first run to save temporary processed netcdf files\n",
    "skip_ncgen_temp = True\n",
    "\n",
    "path_station = 'HEROES'\n",
    "station_folder = {\n",
    "    'Baylling': 'Weather station data at Baylling School (Trashiyangtse)',\n",
    "    'Bjishong': 'Weather station data at Bjishong School (Gasa, Damji)',\n",
    "    'Damthang': 'Weather station data at Damthang School (Haa) ',\n",
    "    'Galing': 'Weather station data at Galing School (Trashigang)',\n",
    "    'Gyelpozhing': 'Weather station data at Gyelpozhing School (Mongar)',\n",
    "    'Jakar': 'Weather station data at Jakar School (Bumthang)',\n",
    "    'Jomotsangkha': 'Weather station data at Jomotsangkha School (Samdrupjongkhar)',\n",
    "    'Khasadrapchu': 'Weather station data at Khasadrapchu School (Thimphu)',\n",
    "    'Nobding': 'Weather station data at Nobding School (Wangdiphodrang)',\n",
    "    'Pelrithang': 'Weather station data at Pelrithang School (Gelephu)',\n",
    "    'Sakteng': 'Weather station data at Sakteng School (Trashigang)',\n",
    "    'Tengmachu': 'Weather station data at Tangmachu School (Lhuntse)',\n",
    "    'Tendruk': 'Weather station data at Tendruk School (Samtse)',\n",
    "    'Yurung': 'Weather station data at Yurung School (Samtse)',\n",
    "}\n",
    "station_lonlat = {\n",
    "    'Baylling': [91.5035, 27.6213],\n",
    "    'Bjishong': [89.7267, 27.809],\n",
    "    'Damthang': [89.2033, 27.4342],\n",
    "    'Galing': [91.6405, 27.339],\n",
    "    'Gyelpozhing': [91.1935, 27.2369],\n",
    "    'Jakar': [90.7376, 27.5458],\n",
    "    'Jomotsangkha': [92.1097, 26.8962],\n",
    "    'Khasadrapchu': [89.5803, 27.3855],\n",
    "    'Nobding': [90.1504, 27.5469],\n",
    "    'Pelrithang': [90.4913, 26.9021],\n",
    "    'Sakteng': [91.923, 27.4031],\n",
    "    'Tengmachu': [91.1914, 27.5922],\n",
    "    'Tendruk': [88.8722, 27.1211],\n",
    "    'Yurung': [91.348, 27.0215],\n",
    "}\n",
    "\n",
    "bounds = [88.6875, 25.5, 93.375, 28.625] # covers the basin over Bhutan\n",
    "bounds_bhutan = [88.6875, 26.6875, 92.125, 28.25] # covers Bhutan\n",
    "date_range = ['2015-01-01','2019-12-31'] # for statistics\n",
    "\n",
    "path_data = '/rcfs/projects/sage/hydro-modeling/datasets'\n",
    "#path_data = 'datasets' # if there is a symlink created for the absolute path above\n",
    "list_models = ['ERA5', 'ERA5-L', 'MSWX-P', 'GSWP3-W5E5', 'HMA', 'GMFD'] # a list of model variables\n",
    "years_load = np.arange(2015, 2021) # for loading datasets\n",
    "path_temps, path_figs = 'tempfiles', 'figures'\n",
    "os.makedirs(path_temps, exist_ok = True)\n",
    "os.makedirs(path_figs, exist_ok = True)\n",
    "\n",
    "vars_list = {\n",
    "    'ERA5': [\n",
    "        't2m', # 2 metre temperature\n",
    "        'tp', # Total precipitation\n",
    "    ],\n",
    "    'ERA5-L': [\n",
    "        't2m', # 2 metre temperature\n",
    "        'tp', # Total precipitation\n",
    "    ],\n",
    "    'MSWX-P': [\n",
    "        'air_temperature',\n",
    "        'air_temperature_max',\n",
    "        'air_temperature_min',\n",
    "        'precipitation',\n",
    "    ],\n",
    "    'GSWP3-W5E5': [\n",
    "        'pr', # Precipitation\n",
    "        'tas', # Near-Surface Air Temperature\n",
    "        'tasmax', # Near-Surface Air Temperature\n",
    "        'tasmin', # Near-Surface Air Temperature\n",
    "    ],\n",
    "    'HMA': [\n",
    "        'P_CHP', # Downscaled CHIRPS total precipitation\n",
    "        'P_ECMWF', # Downscaled ECMWF total precipitation\n",
    "        'Tair', # Downscaled ECMWF near-surface temperature\n",
    "    ],\n",
    "    'GMFD': [\n",
    "        'prcp', # Precipitation\n",
    "        'tas', # Air temperature\n",
    "        'tmax', # Maximum air temperature\n",
    "        'tmin', # Minimum air temperature\n",
    "    ],\n",
    "}\n",
    "key_prcp = {'ERA5': ['tp'], 'ERA5-L': ['tp'], 'MSWX-P': ['precipitation'], 'GSWP3-W5E5': ['pr'], 'HMA': ['P_CHP', 'P_ECMWF'], 'GMFD': ['prcp']} # preipitation variables in each model\n",
    "key_temp = {'ERA5': 't2m', 'ERA5-L': 't2m', 'HMA': 'Tair'} # sub-daily temperature variables in each model\n",
    "key_tmin = {'MSWX-P': 'air_temperature_min', 'GSWP3-W5E5': 'tasmin', 'GMFD': 'tmin'} # daily minimum temperature variables in each model\n",
    "key_tmax = {'MSWX-P': 'air_temperature_max', 'GSWP3-W5E5': 'tasmax', 'GMFD': 'tmax'} # daily maximum temperature variables in each model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_df = {}\n",
    "for key, folder in station_folder.items():\n",
    "    df = pd.read_csv(glob.glob(os.path.join(path_station, folder, 'data', '*_data_monthly.csv'))[0])\n",
    "    df.index = pd.to_datetime(df['Month'].astype(str) + '/1/' + df['Year'].astype(str))\n",
    "    dict_df[key] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_ds = []\n",
    "if not skip_ncgen_prcp or not skip_ncgen_temp:\n",
    "    for m in list_models:\n",
    "        ds = functions.read_datasets(path = path_data, model = m, years = years_load, bounds = bounds, vars_list = vars_list[m], save_nc = None) # load datasets to process\n",
    "        for key in key_prcp[m]: ds[key] = functions.convert_precipitation(da_prcp = ds[key], model = m) # convert a pricipitation variable into total precipitation per step\n",
    "        if m in key_temp.keys(): ds[key_temp[m]] = functions.convert_temperature(da_temp = ds[key_temp[m]], model = m) # convert a sub-daily temperature variable into degreeC\n",
    "        if m in key_tmin.keys(): ds[key_tmin[m]] = functions.convert_temperature(da_temp = ds[key_tmin[m]], model = m) # convert a daily minimum temperature variable into degreeC\n",
    "        if m in key_tmax.keys(): ds[key_tmax[m]] = functions.convert_temperature(da_temp = ds[key_tmax[m]], model = m) # convert a daily maximum temperature variable into degreeC\n",
    "        list_ds.append(ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_name_prcp, list_da_mprcp = [], []\n",
    "if not skip_ncgen_prcp:\n",
    "    for i, m in enumerate(list_models):\n",
    "        for key in key_prcp[m]:\n",
    "            list_name_prcp.append(f'{functions.models_name[m]}: \"{key}\"')\n",
    "            da = list_ds[i][key].resample(time = '1MS', skipna = False).sum() # Monthly Precipitation Total\n",
    "            date_min = max(pd.Timestamp(date_range[0]), pd.Timestamp(da['time'][0].values))\n",
    "            date_max = min(pd.Timestamp(date_range[-1]), pd.Timestamp(da['time'][-1].values))\n",
    "            da_mprcp = da.sel(time = pd.date_range(date_min, date_max, freq = '1MS')).load()\n",
    "            da_mprcp.to_netcdf(os.path.join(path_temps, f'_ncgen_HEROES_prcp_{m}_{key}.nc'))\n",
    "            list_da_mprcp.append(da_mprcp)\n",
    "else:\n",
    "    for i, m in enumerate(list_models):\n",
    "        for key in key_prcp[m]:\n",
    "            list_name_prcp.append(f'{functions.models_name[m]}: \"{key}\"')\n",
    "            list_da_mprcp.append(xr.load_dataarray(os.path.join(path_temps, f'_ncgen_HEROES_prcp_{m}_{key}.nc'), decode_coords = 'all'))\n",
    "\n",
    "list_name_temp, list_da_mtmean, list_da_mtmin, list_da_mtmax = [], [], [], []\n",
    "if not skip_ncgen_temp:\n",
    "    for i, m in enumerate(list_models):\n",
    "        if m in key_temp.keys():\n",
    "            list_name_temp.append(f'{functions.models_name[m]}: \"{key_temp[m]}\"')\n",
    "            da = list_ds[i][key_temp[m]].resample(time = '1MS', skipna = False).mean() # Monthly Mean if the sub-hourly data are available\n",
    "        elif m in key_tmin.keys() and m in key_tmax.keys():\n",
    "            list_name_temp.append(f'{functions.models_name[m]}: \"{key_tmin[m]}, {key_tmax[m]}\"')\n",
    "            da = (list_ds[i][key_tmin[m]] + list_ds[i][key_tmax[m]]) / 2  # Average of Tmax and Tmin if the sub-hourly data are unavailable, similar to the paper\n",
    "            da = da.resample(time = '1MS', skipna = False).mean() # Monthly Mean\n",
    "            da.name = key_tmin[m][0].replace('min','mean_by_minmax')\n",
    "        date_min = max(pd.Timestamp(date_range[0]), pd.Timestamp(da['time'][0].values))\n",
    "        date_max = min(pd.Timestamp(date_range[-1]), pd.Timestamp(da['time'][-1].values))\n",
    "        da_mtmean = da.sel(time = pd.date_range(date_min, date_max, freq = '1MS')).load()\n",
    "        da_mtmean.to_netcdf(os.path.join(path_temps, f'_ncgen_HEROES_tmean_{m}.nc'))\n",
    "        list_da_mtmean.append(da_mtmean)\n",
    "\n",
    "        if m in key_temp.keys(): da_min = list_ds[i][key_temp[m]].resample(time = '1D', skipna = False).min() # Daily Minimum Temperature\n",
    "        elif m in key_tmin.keys() and m in key_tmax.keys(): da_min = list_ds[i][key_tmin[m]]\n",
    "        date_min = max(pd.Timestamp(date_range[0]), pd.Timestamp(da_min['time'][0].values))\n",
    "        date_max = min(pd.Timestamp(date_range[-1]), pd.Timestamp(da_min['time'][-1].values))\n",
    "        da_mtmin = da_min.sel(time = pd.date_range(date_min, date_max, freq = '1D')).resample(time = '1MS', skipna = False).min() # Monthly Min\n",
    "        da_mtmin.to_netcdf(os.path.join(path_temps, f'_ncgen_HEROES_tmin_{m}.nc'))\n",
    "        list_da_mtmin.append(da_mtmin)\n",
    "        \n",
    "        if m in key_temp.keys(): da_max = list_ds[i][key_temp[m]].resample(time = '1D', skipna = False).max() # Daily Maximum Temperature\n",
    "        elif m in key_tmin.keys() and m in key_tmax.keys(): da_max = list_ds[i][key_tmax[m]]\n",
    "        date_min = max(pd.Timestamp(date_range[0]), pd.Timestamp(da_max['time'][0].values))\n",
    "        date_max = min(pd.Timestamp(date_range[-1]), pd.Timestamp(da_max['time'][-1].values))\n",
    "        da_mtmax = da_max.sel(time = pd.date_range(date_min, date_max, freq = '1D')).resample(time = '1MS', skipna = False).max() # Monthly Max\n",
    "        da_mtmax.to_netcdf(os.path.join(path_temps, f'_ncgen_HEROES_tmax_{m}.nc'))\n",
    "        list_da_mtmax.append(da_mtmax)\n",
    "else:\n",
    "    for i, m in enumerate(list_models):\n",
    "        if m in key_temp.keys(): list_name_temp.append(f'{functions.models_name[m]}: \"{key_temp[m]}\"')\n",
    "        elif m in key_tmin.keys() and m in key_tmax.keys(): list_name_temp.append(f'{functions.models_name[m]}: \"{key_tmin[m]}, {key_tmax[m]}\"')\n",
    "        list_da_mtmean.append(xr.load_dataarray(os.path.join(path_temps, f'_ncgen_HEROES_tmean_{m}.nc'), decode_coords = 'all'))\n",
    "        list_da_mtmin.append(xr.load_dataarray(os.path.join(path_temps, f'_ncgen_HEROES_tmin_{m}.nc'), decode_coords = 'all'))\n",
    "        list_da_mtmax.append(xr.load_dataarray(os.path.join(path_temps, f'_ncgen_HEROES_tmax_{m}.nc'), decode_coords = 'all'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Plot Properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "shp_file = os.path.join('shapefiles', 'btn_admbnda_adm1_bnlc_20201026.shp') # Bhutan Administrative Boundaries (Level 1)\n",
    "shp_file_mask = os.path.join('shapefiles', 'gridmask_bhutan.shp') # Polygon masking outside of Bhutan\n",
    "\n",
    "baseMap = cartopy.io.img_tiles.GoogleTiles(style = 'satellite')\n",
    "\n",
    "shape_feature = cartopy.feature.ShapelyFeature(cartopy.io.shapereader.Reader(shp_file).geometries(), crs = cartopy.crs.PlateCarree(), facecolor = 'none', linewidth = 0.5) # Bhutan Administrative Boundaries (Level 1)\n",
    "shape_feature_mask = cartopy.feature.ShapelyFeature(cartopy.io.shapereader.Reader(shp_file_mask).geometries(), crs = cartopy.crs.PlateCarree(), facecolor = 'w', linewidth = 0.5) # masking outside of Bhutan\n",
    "polygon_feature = shapely.unary_union(shapely.geometry.MultiPolygon(shape_feature.geometries())) # applying statistics only for Bhutan\n",
    "\n",
    "# function to select the nearest grid\n",
    "def select_da_nearest(da, lonlat):\n",
    "    if 'lon' in da.dims and 'lat' in da.dims: x, y = 'lon', 'lat'\n",
    "    elif 'longitude' in da.dims and 'latitude' in da.dims: x, y = 'longitude', 'latitude'\n",
    "    elif 'x' in da.dims and 'y' in da.dims: x, y = 'x', 'y'\n",
    "    else: return da\n",
    "    return da.sel({x: lonlat[0], y: lonlat[1]}, method = 'nearest')\n",
    "\n",
    "# function to plot a timeseries of variable\n",
    "def plot_obs(df, df_name, list_da_sel, list_name, title = '', xlabel = '', ylabel = '', xlim = [None, None], ylim = [None, None], savefig = None, show = None):\n",
    "    fig, ax = plt.subplots(figsize = (14, 4))\n",
    "    for i, da_sel in enumerate(list_da_sel): ax.plot(da_sel['time'], da_sel, label = list_name[i])\n",
    "    ax.plot(df.index, df, 'r-o', label = df_name)\n",
    "    ax.set_title(title)\n",
    "    ax.set_xlabel(xlabel); ax.set_ylabel(ylabel)\n",
    "    ax.set_xlim(xlim); ax.set_ylim(ylim)\n",
    "    ax.grid()\n",
    "\n",
    "    fig.subplots_adjust(bottom = 0.15)\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    fig.legend(handles, labels, ncol = 4, loc = 'lower center', bbox_to_anchor = (0, 0.01, 1, 0.02))\n",
    "    fig.tight_layout(rect = [0, 0.15, 1, 1])\n",
    "\n",
    "    if savefig is not None: plt.savefig(savefig, bbox_inches = 'tight')\n",
    "    if show is not None: plt.show()\n",
    "    return\n",
    "\n",
    "# function to create a masked dataarray for statistics\n",
    "def create_da_masked(da, polygon = None):\n",
    "    if polygon is None: return da\n",
    "    if 'lon' in da.dims and 'lat' in da.dims: x, y = 'lon', 'lat'\n",
    "    elif 'longitude' in da.dims and 'latitude' in da.dims: x, y = 'longitude', 'latitude'\n",
    "    elif 'x' in da.dims and 'y' in da.dims: x, y = 'x', 'y'\n",
    "    else: return da\n",
    "    xs, ys = np.meshgrid(da[x], da[y])\n",
    "    xys = np.stack([xs.flatten(), ys.flatten()], axis = -1)\n",
    "    mask = shapely.contains_xy(polygon, xys)\n",
    "    mask = mask.reshape([len(da[y]), len(da[x])])\n",
    "    return da.where(mask == True)\n",
    "\n",
    "# function to plot a geospatial map of variable, including observations\n",
    "def plot_models(list_da, list_name, list_shapes, bounds, norm = None, cmap = 'nipy_spectral', suptitle = '', clabel = '', stats = False, stats_polygon = None, stats_unit = 'mm', savefig = None, show = True, obs = None):\n",
    "    nplot = len(list_da)\n",
    "    fig, axes = plt.subplots(nrows = (1 + nplot) // 2, ncols = 2, figsize = (10, 14), subplot_kw = {'projection': baseMap.crs})\n",
    "    fig.suptitle(suptitle)\n",
    "    for i, ax in enumerate(axes.flat):\n",
    "        ax.set_extent([bounds[0], bounds[2], bounds[1], bounds[3]])\n",
    "        fgl = ax.gridlines(crs = cartopy.crs.PlateCarree(), linestyle = '--', alpha = 0.25, draw_labels = True)\n",
    "        fgl.top_labels, fgl.right_labels = False, False\n",
    "        fgl.xlabel_style, fgl.ylabel_style = {'color': 'gray', 'size': 8}, {'color': 'gray', 'size': 8}\n",
    "\n",
    "        if i < nplot:\n",
    "            fda = list_da[i].plot(ax = ax, transform = cartopy.crs.PlateCarree(), norm = norm, cmap = cmap, add_colorbar = False)\n",
    "            if obs is not None: ax.scatter(obs[:, 0], obs[:, 1], c = obs[:, 2], transform = cartopy.crs.PlateCarree(), norm = norm, cmap = cmap, edgecolors = 'k')\n",
    "            \n",
    "            if stats == True or stats_polygon is not None: # applying statistics only for Bhutan\n",
    "                da_masked = create_da_masked(da = list_da[i], polygon = stats_polygon)\n",
    "                stats = 'Mean: {:.0f} {}, Min: {:.0f} {}, Max: {:.0f} {}'.format(np.nanmean(da_masked), stats_unit, np.nanmin(da_masked), stats_unit, np.nanmax(da_masked), stats_unit)\n",
    "                ax.set_title(list_name[i] + '\\n' + stats)\n",
    "            else: ax.set_title(list_name[i])\n",
    "        else: ax.add_image(baseMap, 10)\n",
    "        for s in list_shapes: ax.add_feature(s)\n",
    "        \n",
    "    fig.subplots_adjust(bottom = 0.05)\n",
    "    cax = fig.add_axes([0.2, 0.025, 0.6, 0.02])\n",
    "    fig.colorbar(fda, label = clabel, cax = cax, orientation = 'horizontal')\n",
    "    fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
    "\n",
    "    if savefig is not None: plt.savefig(savefig, bbox_inches = 'tight')\n",
    "    if show is not None: plt.show()\n",
    "    return\n",
    "\n",
    "# colormaps\n",
    "precip_11lev = np.array([\n",
    "    [255, 255, 255],\n",
    "    [237, 250, 194],\n",
    "    [205, 255, 205],\n",
    "    [153, 240, 178],\n",
    "    [83, 189, 159],\n",
    "    [50, 166, 150],\n",
    "    [50, 150, 180],\n",
    "    [5, 112, 176],\n",
    "    [5, 80, 140],\n",
    "    [10, 31, 150],\n",
    "    [44, 2, 70],\n",
    "    [106, 44, 90],\n",
    "])\n",
    "precip_11lev = precip_11lev / 255\n",
    "\n",
    "prcp_bounds = [0, 1, 2, 5, 10, 20, 200, 300, 500, 1000]\n",
    "prcp_norm = matplotlib.colors.BoundaryNorm(boundaries = prcp_bounds, ncolors = 13, extend = 'max')\n",
    "\n",
    "prcp_bounds_tminmax = np.linspace(-15, 30, 10, endpoint = True)\n",
    "prcp_norm_tminmax = matplotlib.colors.BoundaryNorm(boundaries = prcp_bounds_tminmax, ncolors = 256, extend = 'both')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Precipitation Total at Stations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Monthly Precipitation Total at Stations\n",
    "for key, df in dict_df.items():\n",
    "    if 'Rainfall' in df.columns:\n",
    "        df.loc[df['Rainfall'] == '#VALUE!', 'Rainfall'] = np.nan\n",
    "        plot_obs(df = df['Rainfall'].astype(float), df_name = '{} School'.format(key),\n",
    "                list_da_sel = [select_da_nearest(da, station_lonlat[key]) for da in list_da_mprcp], list_name = list_name_prcp,\n",
    "                title = 'Monthly Precipitation Total at {} School (Lon: {:.3f} deg, Lat: {:.3f} deg)'.format(key, station_lonlat[key][0], station_lonlat[key][1]), xlabel = 'Time', ylabel = 'Precipitation [mm]', xlim = [None, None], ylim = [None, None],\n",
    "                savefig = os.path.join(path_figs, f'HEROES-PRCP-STATION-{key}.png'), show = None)\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Mean Temperature at Stations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Monthly Mean Temperature at Stations\n",
    "for key, df in dict_df.items():\n",
    "    if 'Tmean' in df.columns:\n",
    "        df.loc[df['Tmean'] == '#VALUE!', 'Tmean'] = np.nan\n",
    "        plot_obs(df = df['Tmean'].astype(float), df_name = '{} School'.format(key),\n",
    "                list_da_sel = [select_da_nearest(da, station_lonlat[key]) for da in list_da_mtmean], list_name = list_name_temp,\n",
    "                title = 'Monthly Mean Temperature at {} School (Lon: {:.3f} deg, Lat: {:.3f} deg)'.format(key, station_lonlat[key][0], station_lonlat[key][1]), xlabel = 'Time', ylabel = 'Temperature [degC]', xlim = [None, None], ylim = [None, None],\n",
    "                savefig = os.path.join(path_figs, f'HEROES-TMEAN-STATION-{key}.png'), show = None)\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Minimum Temperature at Stations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Monthly Minimum Temperature at Stations\n",
    "for key, df in dict_df.items():\n",
    "    if 'Tmin' in df.columns:\n",
    "        df.loc[df['Tmin'] == '#VALUE!', 'Tmin'] = np.nan\n",
    "        plot_obs(df = df['Tmin'].astype(float), df_name = '{} School'.format(key),\n",
    "                list_da_sel = [select_da_nearest(da, station_lonlat[key]) for da in list_da_mtmin], list_name = list_name_temp,\n",
    "                title = 'Monthly Minimum Temperature at {} School (Lon: {:.3f} deg, Lat: {:.3f} deg)'.format(key, station_lonlat[key][0], station_lonlat[key][1]), xlabel = 'Time', ylabel = 'Temperature [degC]', xlim = [None, None], ylim = [None, None],\n",
    "                savefig = os.path.join(path_figs, f'HEROES-TMIN-STATION-{key}.png'), show = None)\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Maximum Temperature at Stations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Monthly Maximum Temperature at Stations\n",
    "for key, df in dict_df.items():\n",
    "    if 'Tmax' in df.columns:\n",
    "        df.loc[df['Tmax'] == '#VALUE!', 'Tmax'] = np.nan\n",
    "        plot_obs(df = df['Tmax'].astype(float), df_name = '{} School'.format(key),\n",
    "                list_da_sel = [select_da_nearest(da, station_lonlat[key]) for da in list_da_mtmax], list_name = list_name_temp,\n",
    "                title = 'Monthly Maximum Temperature at {} School (Lon: {:.3f} deg, Lat: {:.3f} deg)'.format(key, station_lonlat[key][0], station_lonlat[key][1]), xlabel = 'Time', ylabel = 'Temperature [degC]', xlim = [None, None], ylim = [None, None],\n",
    "                savefig = os.path.join(path_figs, f'HEROES-TMAX-STATION-{key}.png'), show = None)\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Precipitation Total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/265708999.py:76: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n"
     ]
    }
   ],
   "source": [
    "# Monthly Precipitation Total (Bhutan) w/ statistics\n",
    "for month in pd.date_range(date_range[0], date_range[1], freq = '1MS'):\n",
    "    obs = []\n",
    "    for df in dict_df.values():\n",
    "        try: obs.append(float(df['Rainfall'].loc[month]))\n",
    "        except: obs.append(np.inf)\n",
    "    obs = np.concatenate([np.stack(list(station_lonlat.values())), np.array(obs)[:, None]], axis = 1)\n",
    "    list_da = []\n",
    "    for da in list_da_mprcp:\n",
    "        if month in da['time']: list_da.append(da.sel(time = month))\n",
    "        else: list_da.append(da.isel(time = 0) * np.inf)\n",
    "    \n",
    "    plot_models(list_da = list_da, list_name = list_name_prcp,\n",
    "                list_shapes = [shape_feature, shape_feature_mask], bounds = bounds_bhutan,\n",
    "                norm = prcp_norm, cmap = matplotlib.colors.ListedColormap(precip_11lev), suptitle = month.strftime('%Y-%m'), clabel = 'Monthly Precipitation Total [mm]',\n",
    "                stats = True, stats_polygon = polygon_feature, stats_unit = 'mm',\n",
    "                savefig = os.path.join(path_figs, f'HEROES-PRCP-{month.strftime(\"%Y-%m\")}-Bhutan.png'), show = None,\n",
    "                obs = obs)\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modify Plot Properties for Temperature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to plot a geospatial map of variable, including observations\n",
    "def plot_models(list_da, list_name, list_shapes, bounds, norm = None, cmap = 'nipy_spectral', suptitle = '', clabel = '', stats = False, stats_polygon = None, stats_unit = 'mm', savefig = None, show = True, obs = None):\n",
    "    nplot = len(list_da)\n",
    "    fig, axes = plt.subplots(nrows = (1 + nplot) // 2, ncols = 2, figsize = (10, 10), subplot_kw = {'projection': baseMap.crs})\n",
    "    fig.suptitle(suptitle)\n",
    "    for i, ax in enumerate(axes.flat):\n",
    "        ax.set_extent([bounds[0], bounds[2], bounds[1], bounds[3]])\n",
    "        fgl = ax.gridlines(crs = cartopy.crs.PlateCarree(), linestyle = '--', alpha = 0.25, draw_labels = True)\n",
    "        fgl.top_labels, fgl.right_labels = False, False\n",
    "        fgl.xlabel_style, fgl.ylabel_style = {'color': 'gray', 'size': 8}, {'color': 'gray', 'size': 8}\n",
    "\n",
    "        if i < nplot:\n",
    "            fda = list_da[i].plot(ax = ax, transform = cartopy.crs.PlateCarree(), norm = norm, cmap = cmap, add_colorbar = False)\n",
    "            if obs is not None: ax.scatter(obs[:, 0], obs[:, 1], c = obs[:, 2], transform = cartopy.crs.PlateCarree(), norm = norm, cmap = cmap, edgecolors = 'k')\n",
    "            \n",
    "            if stats == True or stats_polygon is not None: # applying statistics only for Bhutan\n",
    "                da_masked = create_da_masked(da = list_da[i], polygon = stats_polygon)\n",
    "                stats = 'Mean: {:.0f} {}, Min: {:.0f} {}, Max: {:.0f} {}'.format(np.nanmean(da_masked), stats_unit, np.nanmin(da_masked), stats_unit, np.nanmax(da_masked), stats_unit)\n",
    "                ax.set_title(list_name[i] + '\\n' + stats)\n",
    "            else: ax.set_title(list_name[i])\n",
    "        else: ax.add_image(baseMap, 10)\n",
    "        for s in list_shapes: ax.add_feature(s)\n",
    "        \n",
    "    fig.subplots_adjust(bottom = 0.05)\n",
    "    cax = fig.add_axes([0.2, 0.025, 0.6, 0.02])\n",
    "    fig.colorbar(fda, label = clabel, cax = cax, orientation = 'horizontal')\n",
    "    fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
    "\n",
    "    if savefig is not None: plt.savefig(savefig, bbox_inches = 'tight')\n",
    "    if show is not None: plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Mean Temperature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n"
     ]
    }
   ],
   "source": [
    "# Monthly Mean Temperature (Bhutan) w/ statistics\n",
    "for month in pd.date_range(date_range[0], date_range[1], freq = '1MS'):\n",
    "    obs = []\n",
    "    for df in dict_df.values():\n",
    "        try: obs.append(float(df['Tmean'].loc[month]))\n",
    "        except: obs.append(np.inf)\n",
    "    obs = np.concatenate([np.stack(list(station_lonlat.values())), np.array(obs)[:, None]], axis = 1)\n",
    "    list_da = []\n",
    "    for da in list_da_mtmean:\n",
    "        if month in da['time']: list_da.append(da.sel(time = month))\n",
    "        else: list_da.append(da.isel(time = 0) * np.inf)\n",
    "    \n",
    "    plot_models(list_da = list_da, list_name = list_name_temp,\n",
    "                list_shapes = [shape_feature, shape_feature_mask], bounds = bounds_bhutan,\n",
    "                norm = prcp_norm_tminmax, cmap = 'nipy_spectral', suptitle = month.strftime('%Y-%m'), clabel = 'Monthly Mean Temperature [degC]',\n",
    "                stats = True, stats_polygon = polygon_feature, stats_unit = 'degC',\n",
    "                savefig = os.path.join(path_figs, f'HEROES-TMEAN-{month.strftime(\"%Y-%m\")}-Bhutan.png'), show = None,\n",
    "                obs = obs)\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Minimum Temperature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n"
     ]
    }
   ],
   "source": [
    "# Monthly Minimum Temperature (Bhutan) w/ statistics\n",
    "for month in pd.date_range(date_range[0], date_range[1], freq = '1MS'):\n",
    "    obs = []\n",
    "    for df in dict_df.values():\n",
    "        try: obs.append(float(df['Tmin'].loc[month]))\n",
    "        except: obs.append(np.inf)\n",
    "    obs = np.concatenate([np.stack(list(station_lonlat.values())), np.array(obs)[:, None]], axis = 1)\n",
    "    list_da = []\n",
    "    for da in list_da_mtmin:\n",
    "        if month in da['time']: list_da.append(da.sel(time = month))\n",
    "        else: list_da.append(da.isel(time = 0) * np.inf)\n",
    "    \n",
    "    plot_models(list_da = list_da, list_name = list_name_temp,\n",
    "                list_shapes = [shape_feature, shape_feature_mask], bounds = bounds_bhutan,\n",
    "                norm = prcp_norm_tminmax, cmap = 'nipy_spectral', suptitle = month.strftime('%Y-%m'), clabel = 'Monthly Minimum Temperature [degC]',\n",
    "                stats = True, stats_polygon = polygon_feature, stats_unit = 'degC',\n",
    "                savefig = os.path.join(path_figs, f'HEROES-TMIN-{month.strftime(\"%Y-%m\")}-Bhutan.png'), show = None,\n",
    "                obs = obs)\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monthly Maximum Temperature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/people/sony061/.conda/envs/geospatial/lib/python3.9/site-packages/numpy/core/fromnumeric.py:88: RuntimeWarning: invalid value encountered in reduce\n",
      "  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)\n",
      "/tmp/ipykernel_7881/3200789245.py:27: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout(rect = [0, 0.05, 1, 1])\n"
     ]
    }
   ],
   "source": [
    "# Monthly Maximum Temperature (Bhutan) w/ statistics\n",
    "for month in pd.date_range(date_range[0], date_range[1], freq = '1MS'):\n",
    "    obs = []\n",
    "    for df in dict_df.values():\n",
    "        try: obs.append(float(df['Tmax'].loc[month]))\n",
    "        except: obs.append(np.inf)\n",
    "    obs = np.concatenate([np.stack(list(station_lonlat.values())), np.array(obs)[:, None]], axis = 1)\n",
    "    list_da = []\n",
    "    for da in list_da_mtmax:\n",
    "        if month in da['time']: list_da.append(da.sel(time = month))\n",
    "        else: list_da.append(da.isel(time = 0) * np.inf)\n",
    "    \n",
    "    plot_models(list_da = list_da, list_name = list_name_temp,\n",
    "                list_shapes = [shape_feature, shape_feature_mask], bounds = bounds_bhutan,\n",
    "                norm = prcp_norm_tminmax, cmap = 'nipy_spectral', suptitle = month.strftime('%Y-%m'), clabel = 'Monthly Maximum Temperature [degC]',\n",
    "                stats = True, stats_polygon = polygon_feature, stats_unit = 'degC',\n",
    "                savefig = os.path.join(path_figs, f'HEROES-TMAX-{month.strftime(\"%Y-%m\")}-Bhutan.png'), show = None,\n",
    "                obs = obs)\n",
    "    plt.close('all')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
